### PR TITLE
Refactor data termination

### PR DESCRIPTION
**Pull request by theghost5800**
List blobs only once and check the whole list of deleted spaces to match
specified filter. Add temporary cleaners of orphaned data which will not
be deleted from Data Termination cleaner

JIRA:LMCROSSITXSADEPLOY-2248

### CODE REVIEW

File: multiapps-controller-persistence/src/main/java/org/cloudfoundry/multiapps/controller/persistence/query/providers/SqlFileQueryProvider.java [hunk 1]
```diff
@@ -226,14 +226,18 @@ protected SqlFileQueryProvider(String tableName, DataSourceDialect dataSourceDia
...
             PreparedStatement statement = null;
             try {
                 statement = connection.prepareStatement(getQuery(DELETE_FILES_BY_SPACE));
-                statement.setString(1, space);
-                int deletedFiles = statement.executeUpdate();
-                logger.debug(MessageFormat.format(Messages.DELETED_0_FILES_WITH_SPACE_1, deletedFiles, space));
+                addSpacesAsBatches(statement, spaces);
+                int[] batchResults = statement.executeBatch();
+                int deletedFiles = 0;
+                for (int batchResult : batchResults) {
```
Comment (IvanBorislavovDimitrov) [inline multiapps-controller-persistence/src/main/java/org/cloudfoundry/multiapps/controller/persistence/query/providers/SqlFileQueryProvider.java:237]: Arrays.stream(batchResults).sum();

File: multiapps-controller-persistence/src/main/java/org/cloudfoundry/multiapps/controller/persistence/query/providers/SqlFileQueryProvider.java [hunk 2]
```diff
@@ -226,14 +226,18 @@ protected SqlFileQueryProvider(String tableName, DataSourceDialect dataSourceDia
...
+                    deletedFiles += batchResult;
+                }
```
Comment (radoslav-d) [inline multiapps-controller-persistence/src/main/java/org/cloudfoundry/multiapps/controller/persistence/query/providers/SqlFileQueryProvider.java:239]: Could be replaced by this one-liner: `int deletedFiles = IntStream.of(batchResult).sum();`

File: multiapps-controller-persistence/src/main/java/org/cloudfoundry/multiapps/controller/persistence/services/FileSystemFileStorage.java [hunk 3]
```diff
@@ -82,14 +82,9 @@ public void deleteFile(String id, String space) throws FileStorageException {
...
-        File spaceDirectory = getSpaceDirectory(space).toFile();
-        try {
-            if (spaceDirectory.exists()) {
-                FileUtils.deleteDirectory(spaceDirectory);
-            }
-        } catch (IOException e) {
-            throw new FileStorageException(MessageFormat.format(Messages.ERROR_DELETING_DIRECTORY, spaceDirectory), e);
+    public void deleteFilesBySpaces(List<String> spaces) throws FileStorageException {
+        for (String space : spaces) {
+            deleteFilesBySpace(space);
```
Comment (IvanBorislavovDimitrov) [inline multiapps-controller-persistence/src/main/java/org/cloudfoundry/multiapps/controller/persistence/services/FileSystemFileStorage.java:87]: If this method throws an exception, some files for spaces will not be deleted. Do you think it is worth wrapping it into a try-catch?

File: multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/jobs/OrphanedDataCleaner.java [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.http.HttpStatus;
+
+import com.sap.cloudfoundry.client.facade.CloudControllerClient;
+import com.sap.cloudfoundry.client.facade.CloudControllerClientImpl;
+import com.sap.cloudfoundry.client.facade.CloudCredentials;
+import com.sap.cloudfoundry.client.facade.CloudOperationException;
+
+public abstract class OrphanedDataCleaner<T extends AuditableConfiguration> implements Cleaner {
```
Comment (radoslav-d) [inline multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/jobs/OrphanedDataCleaner.java:20]: really nice approach!

File: multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/jobs/ConfigurationEntriesCleaner.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+import org.cloudfoundry.multiapps.controller.core.util.ApplicationConfiguration;
+import org.cloudfoundry.multiapps.controller.persistence.model.ConfigurationEntry;
+import org.cloudfoundry.multiapps.controller.persistence.services.ConfigurationEntryService;
+import org.springframework.core.annotation.Order;
+
+@Named()
+@Order(40)
+public class ConfigurationEntriesCleaner extends OrphanedDataCleaner<ConfigurationEntry> {
+
+    private ConfigurationEntryService configurationEntryService;
```
Comment (IvanBorislavovDimitrov) [inline multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/jobs/ConfigurationEntriesCleaner.java:17]: Make it final

File: multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/jobs/ConfigurationEntriesCleaner.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    @Inject
+    public ConfigurationEntriesCleaner(ApplicationConfiguration applicationConfiguration,
+                                       ConfigurationEntryService configurationEntryService) {
+        super(applicationConfiguration);
+        this.configurationEntryService = configurationEntryService;
+    }
+
+    @Override
+    protected String getStartCleanupLogMessage() {
+        return "Deleting orphaned configuration entries...";
```
Comment (IvanBorislavovDimitrov) [inline multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/jobs/ConfigurationEntriesCleaner.java:28]: Can you extract these messages in constants?

File: multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/jobs/ConfigurationSubscriptionCleaner.java [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    @Inject
+    public ConfigurationSubscriptionCleaner(ApplicationConfiguration applicationConfiguration,
+                                            ConfigurationSubscriptionService configurationSubscriptionService) {
+        super(applicationConfiguration);
+        this.configurationSubscriptionService = configurationSubscriptionService;
+    }
+
+    @Override
+    protected String getStartCleanupLogMessage() {
+        return "Deleting orphaned configuration subscriptions...";
```
Comment (IvanBorislavovDimitrov) [inline multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/jobs/ConfigurationSubscriptionCleaner.java:28]: Constants?

File: multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/jobs/ConfigurationSubscriptionCleaner.java [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+import org.cloudfoundry.multiapps.controller.core.util.ApplicationConfiguration;
+import org.cloudfoundry.multiapps.controller.persistence.model.ConfigurationSubscription;
+import org.cloudfoundry.multiapps.controller.persistence.services.ConfigurationSubscriptionService;
+import org.springframework.core.annotation.Order;
+
+@Named()
+@Order(40)
+public class ConfigurationSubscriptionCleaner extends OrphanedDataCleaner<ConfigurationSubscription> {
+
+    private ConfigurationSubscriptionService configurationSubscriptionService;
```
Comment (IvanBorislavovDimitrov) [inline multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/jobs/ConfigurationSubscriptionCleaner.java:17]: final?

**Review by radoslav-d — APPROVED** [review]

**Review by IvanBorislavovDimitrov — APPROVED** [review]

**Review by radoslav-d — APPROVED** [review]

**Pull request merged**